search/v2/features_filter.hpp

```diff
@@ -16,17 +21,58 @@ class FeaturesFilter
 public:
   FeaturesFilter();
 
-  FeaturesFilter(unique_ptr<coding::CompressedBitVector> filter, uint32_t threshold);
+  FeaturesFilter(coding::CompressedBitVector const & filter, uint32_t threshold);
+
+  inline void SetFilter(coding::CompressedBitVector const * filter) { m_filter = filter; }
 
-  inline void SetFilter(unique_ptr<coding::CompressedBitVector> filter) { m_filter = move(filter); }
   inline void SetThreshold(uint32_t threshold) { m_threshold = threshold; }
 
   bool NeedToFilter(coding::CompressedBitVector const & features) const;
   unique_ptr<coding::CompressedBitVector> Filter(coding::CompressedBitVector const & cbv) const;
 
 private:
-  unique_ptr<coding::CompressedBitVector> m_filter;
+  coding::CompressedBitVector const * m_filter;
   uint32_t m_threshold;
 };
+
+
+/// CompressedBitVector pointer class that incapsulates
+/// binary operators logic and takes ownership if needed.
+class CBVPtr
```

Comment:
In any case, have you considered to use shared_ptr instead of new custom smart pointer? They have overhead on atomics (~100ns on modern CPUs), but do not require new code.